### PR TITLE
qgsmesh3geometry: Limit geometry to 3d scenes' 2d extent

### PR DESCRIPTION
**Pull request by ptitjano**
## Description

From the main commit message: 

```
This is a follow-up of https://github.com/qgis/QGIS/pull/51304 which
did not handle mesh layers. This commit adds this support.

The meshes displayed in a 3d view do not take into account the scene
2d extent.

This issue is fixed by filtering the faces which are used
to build the Qt3D geometry. By using `mesh.faceIndexesForRectangle(
extent )`, it is possible to keep the faces inside `extent`. This list
is then used to populate the `IndexAttribute` of the qt3d
geometry. This way, only the faces inside the extent are displayed.
```

## Current behavior: The whole mesh is displayed

nocut

## New behavior: Only the relevant part of the mesh is displayed

cut

cc @uclaros @vcloarec @benoitdm-oslandia

### CODE REVIEW

File: src/3d/mesh/qgsmesh3dentity_p.cpp [hunk 1]
```diff
@@ -54,7 +54,7 @@ void QgsMeshDataset3DEntity::buildGeometry()
     return;
 
   Qt3DRender::QGeometryRenderer *mesh = new Qt3DRender::QGeometryRenderer;
-  mesh->setGeometry( new QgsMeshDataset3DGeometry( mTriangularMesh, layer(), mMapSettings.temporalRange(), mMapSettings.origin(), mSymbol.get(), mesh ) );
+  mesh->setGeometry( new QgsMeshDataset3DGeometry( mTriangularMesh, layer(), mMapSettings.temporalRange(), mMapSettings.origin(), mMapSettings.extent(), mSymbol.get(), mesh ) );
```
[the remark this answers is not in the document]
Comment (vcloarec) [reply src/3d/mesh/qgsmesh3dentity_p.cpp:57]: `mTriangularMesh` should be in map coordinates

File: src/3d/mesh/qgsmesh3dgeometry_p.cpp [hunk 2]
```diff
@@ -132,25 +132,25 @@ static QByteArray createDatasetVertexData(
   return bufferBytes;
 }
 
-static QByteArray createIndexData( const QgsTriangularMesh &mesh )
+static QByteArray createIndexData( const QgsTriangularMesh &mesh, const QgsRectangle &extent )
 {
-  const int faces = mesh.triangles().count();
-  const quint32 indices = static_cast<quint32>( 3 * faces );
+  QList<int> facesInExtent = mesh.faceIndexesForRectangle( extent );
```
Comment (vcloarec) [inline src/3d/mesh/qgsmesh3dgeometry_p.cpp:137]: `const` ?

File: src/3d/mesh/qgsmesh3dgeometry_p.cpp [hunk 3]
```diff
@@ -160,36 +160,34 @@ static QByteArray createIndexData( const QgsTriangularMesh &mesh )
   return indexBytes;
 }
 
-static QByteArray createDatasetIndexData( const QgsTriangularMesh &mesh, const QgsMeshDataBlock &mActiveFaceFlagValues )
+static QByteArray createDatasetIndexData( const QgsTriangularMesh &mesh, const QgsMeshDataBlock &mActiveFaceFlagValues, const QgsRectangle &extent )
 {
+  QList<int> facesInExtent = mesh.faceIndexesForRectangle( extent );
```
Comment (vcloarec) [inline src/3d/mesh/qgsmesh3dgeometry_p.cpp:165]: `cons`t ?

**Review by vcloarec — CHANGES_REQUESTED** [review]
well done !

File: src/3d/mesh/qgsmesh3dgeometry_p.cpp (hunk 3 above)
Comment (ptitjano) [reply src/3d/mesh/qgsmesh3dgeometry_p.cpp]: Done.

File: src/3d/mesh/qgsmesh3dgeometry_p.cpp (hunk 2 above)
Comment (ptitjano) [reply src/3d/mesh/qgsmesh3dgeometry_p.cpp]: Done.

File: src/3d/mesh/qgsmesh3dentity_p.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ptitjano) [reply src/3d/mesh/qgsmesh3dentity_p.cpp:57]: Indeed.

**Review by vcloarec — APPROVED** [review]

**Comment by uclaros** [comment]
Nice! Thanks @ptitjano !

**Comment by uclaros** [comment]
IIRC there's a test for 3d rendering a mesh. Could you also add a test case with limiting the extents?

**Comment by ptitjano** [comment]
> IIRC there's a test for 3d rendering a mesh. Could you also add a test case with limiting the extents?

Good idea ! Done.

**Review by benoitdm-oslandia — APPROVED** [review]
LGTM

**Comment by ptitjano** [comment]
@nyalldawson I believe this is ready to be merged. All the requested changes have been made.

**Pull request merged**